File: sdks/csharp/Renamed/Exceptions/RenamedExceptionBase.cs
namespace Renamed.Sdk.Exceptions;

/// <summary>
/// Base exception class for all renamed.to SDK exceptions.
/// </summary>
public class RenamedExceptionBase : Exception
{
    /// <summary>
    /// Error code identifying the type of error.
    /// </summary>
    public string Code { get; }

    /// <summary>
    /// HTTP status code associated with the error, if applicable.
    /// </summary>
    public int? StatusCode { get; }

    /// <summary>
    /// Additional error details from the API response.
    /// </summary>
    public object? Details { get; }

    /// <summary>
    /// Creates a new instance of RenamedExceptionBase.
    /// </summary>
    /// <param name="message">The error message.</param>
    /// <param name="code">The error code.</param>
    /// <param name="statusCode">The HTTP status code, if applicable.</param>
    /// <param name="details">Additional error details.</param>
    public RenamedExceptionBase(
        string message,
        string code,
        int? statusCode = null,
        object? details = null)
        : base(message)
    {
        Code = code;
        StatusCode = statusCode;
        Details = details;
    }

    /// <summary>
    /// Creates a new instance of RenamedExceptionBase with an inner exception.
    /// </summary>
    /// <param name="message">The error message.</param>
    /// <param name="code">The error code.</param>
    /// <param name="innerException">The inner exception.</param>
    /// <param name="statusCode">The HTTP status code, if applicable.</param>
    /// <param name="details">Additional error details.</param>
    public RenamedExceptionBase(
        string message,
        string code,
        Exception innerException,
        int? statusCode = null,
        object? details = null)
        : base(message, innerException)
    {
        Code = code;
        StatusCode = statusCode;
        Details = details;
    }
}

File: sdks/php/src/Exceptions/RenamedExceptionBase.php
<?php

declare(strict_types=1);

namespace Renamed\Exceptions;

use Exception;
use Throwable;

/**
 * Base exception for all renamed.to SDK errors.
 */
class RenamedExceptionBase extends Exception
{
    protected string $errorCode;
    protected ?int $statusCode;
    protected mixed $details;

    public function __construct(
        string $message = 'An error occurred',
        string $errorCode = 'UNKNOWN_ERROR',
        ?int $statusCode = null,
        mixed $details = null,
        ?Throwable $previous = null
    ) {
        parent::__construct($message, 0, $previous);
        $this->errorCode = $errorCode;
        $this->statusCode = $statusCode;
        $this->details = $details;
    }

    public function getErrorCode(): string
    {
        return $this->errorCode;
    }

    public function getStatusCode(): ?int
    {
        return $this->statusCode;
    }

    public function getDetails(): mixed
    {
        return $this->details;
    }

    /**
     * Create appropriate exception from HTTP status code.
     */
    public static function fromHttpStatus(int $status, string $statusText, mixed $payload = null): self
    {
        $message = $statusText;
        if (is_array($payload) && isset($payload['error'])) {
            $message = (string) $payload['error'];
        }

        return match ($status) {
            401 => new AuthenticationException($message),
            402 => new InsufficientCreditsException($message),
            400, 422 => new ValidationException($message, $payload),
            429 => new RateLimitException(
                $message,
                is_array($payload) ? ($payload['retryAfter'] ?? null) : null
            ),
            default => new self($message, 'API_ERROR', $status, $payload),
        };
    }
}
